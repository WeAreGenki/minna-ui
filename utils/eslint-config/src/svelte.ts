/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

interface BlockAttributes {
  type: string;
}

export = {
  plugins: ['svelte3'],
  settings: {
    'svelte3/ignore-styles': (attr: BlockAttributes) =>
      attr.type === 'text/postcss',
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
