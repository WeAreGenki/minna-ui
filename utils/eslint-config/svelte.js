/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 *
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

'use strict';

/** @type {import('./types').ESLintConfig} */
module.exports = {
  plugins: ['svelte3'],
  settings: {
    /**
     * @param {import('./types').BlockAttributes} attr - Tag attributes.
     * @returns {boolean} If style block should be ignored.
     */
    'svelte3/ignore-styles': (attr) => attr.type === 'text/postcss',
    // TODO: Uncomment once false positives are handled better
    // 'svelte3/lint-template': true,
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
