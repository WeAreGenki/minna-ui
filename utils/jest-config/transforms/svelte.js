/**
 * Jest transform for compiling Svelte components into JavaScript.
 *
 * NOTE: `svelte` is not listed as a dependency so developers can use this
 * jest preset for non-svelte projects.
 */

'use strict';

const { compile } = require('svelte/compiler'); // eslint-disable-line import/no-extraneous-dependencies

/** @type {import('../types').JestTransformer} */
exports.process = (src, filename) => {
  // Strip out <style> tags to prevent errors when unable to parse PostCSS etc.
  const re = /<style[^>]*>[\S\s]*?<\/style>/g;
  const normalised = src.replace(re, '');

  const result = compile(normalised, {
    accessors: true, // Easy access to component props for tests
    css: false,
    dev: true,
    filename,
    format: 'cjs',
    preserveWhitespace: true,
  });

  const esInterop =
    'Object.defineProperty(exports, "__esModule", { value: true });';

  return {
    code: result.js.code + esInterop,
    map: result.js.map,
  };
};
