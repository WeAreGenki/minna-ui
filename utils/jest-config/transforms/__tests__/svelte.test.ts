/* eslint-env browser */
/* eslint-disable @typescript-eslint/no-non-null-assertion, security/detect-eval-with-expression, no-eval, global-require */

import fs from 'fs';
import { join } from 'path';
import { promisify } from 'util';
import { process } from '../svelte.js';

// eslint-disable-next-line security/detect-non-literal-fs-filename
const readFile = promisify(fs.readFile);
const sourcePath = join(__dirname, '../../fixtures/TestComponent.svelte');

let source = '';

beforeAll(async () => {
  source = await readFile(sourcePath, 'utf8');
});

describe('Svelte transform', () => {
  it('compiles and mounts a component', () => {
    expect.assertions(2);
    // @ts-expect-error
    const { code } = process(source, sourcePath);
    expect(typeof code).toBe('string');
    const { default: TestComponent } = eval(code);
    const target = document.createElement('div');
    new TestComponent({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });

  it('has access to Svelte component internals when mounted', () => {
    expect.assertions(13);
    // @ts-expect-error
    const { code } = process(source, sourcePath);
    const { default: TestComponent } = eval(code);
    const target = document.createElement('div');
    const component = new TestComponent({ target });

    // svelte component internals
    expect(component.$$).toHaveProperty('fragment');
    expect(component.$$).toHaveProperty('ctx');
    expect(component.$$).toHaveProperty('update');
    expect(component.$$).toHaveProperty('on_mount');
    expect(component.$$).toHaveProperty('on_destroy');
    expect(component.$$).toHaveProperty('before_update');
    expect(component.$$).toHaveProperty('after_update');
    expect(component.$$).toHaveProperty('context');
    expect(component.$$).toHaveProperty('callbacks');

    const name1 = target.querySelector<HTMLDivElement>('#name')!;
    expect(name1.textContent).toBe('Name: Elon Musk');
    const nameReversed1 = target.querySelector<HTMLDivElement>('#nameReversed')!;
    expect(nameReversed1.textContent).toBe('Name: ksuM nolE');
    component.name = 'Vladimir Putin';
    const name2 = target.querySelector<HTMLDivElement>('#name')!;
    expect(name2.textContent).toBe('Name: Vladimir Putin');
    const nameReversed2 = target.querySelector<HTMLDivElement>('#nameReversed')!;
    expect(nameReversed2.textContent).toBe('Name: nituP rimidalV');
  });

  it('mounts components which import ES6 modules', () => {
    expect.assertions(5);
    function wrapper(): void {
      // Use require() instead of process() + eval() so imports are relative
      const TestComponent = require('../../fixtures/TestComponentImports.svelte').default;
      const target = document.createElement('div');
      new TestComponent({ target });
      const name = target.querySelector<HTMLDivElement>('#name')!;
      expect(name.textContent).toBe('Elon Musk');
      const loud = target.querySelector<HTMLDivElement>('#loud')!;
      expect(loud.textContent).toBe('ELON MUSK');
      const quiet = target.querySelector<HTMLDivElement>('#quiet')!;
      expect(quiet.textContent).toBe('elon musk');
      expect(target.innerHTML).toMatchSnapshot();
    }
    expect(wrapper).not.toThrow();
  });
});
