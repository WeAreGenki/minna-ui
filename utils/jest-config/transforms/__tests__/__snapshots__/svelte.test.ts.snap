// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svelte transform compiles and mounts a component 1`] = `
"<div class=\\"wrapper\\"><div id=\\"name\\" class=\\"text\\">test Elon Musk</div>

  <div id=\\"nameReversed\\" class=\\"text\\">test <span>ksuM nolE</span></div></div>"
`;

exports[`Svelte transform mounts components which import ES6 modules 1`] = `"<div id=\\"name\\">Elon Musk</div> <div id=\\"loud\\">ELON MUSK</div> <div id=\\"quiet\\">elon musk</div>"`;
